### PR TITLE
[AP] Partial Legalizer

### PR DESCRIPTION
**Pull request by AlexandreSinger**
This adds the Patial Legalizer block into VPR. A Partial Legalizer takes a partial placement (which likely has many overlapping blocks and is very illegal with blocks in the wrong tiles) and creates a more legal partial placement. These are often called spreaders since their main task is to spread blocks away from each other (the solver tends to put blocks on top of each other).

Added a multi-commodity flow-based spreading partial legalizer based on the work by Darav et al. This algorithm was modified to work within the VPR infrastructure on any hypothetical architecture which can be expressed in VPR.

### CODE REVIEW

File: vpr/src/analytical_place/partial_legalizer.cpp [hunk 1]
```diff
@@ -0,0 +1,1104 @@
...
+        // If there are no overfilled bins, no more work to do.
+        if (overfilled_bins_.empty()) {
+            VTR_LOGV(log_verbosity_ >= 10,
+                     "Flow-Based legalizer has no overfilled tiles. No further spreading needed.\n");
+            break;
+        }
+        // Compute the max movement.
+        double psi = computeMaxMovement(flowBasedIter);
+        // Get the overfilled bins and sort them in increasing order of supply.
+        // We take the manhatten (L1) norm here since we only care about the total
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.cpp:1039]: manhattan

File: vpr/src/analytical_place/partial_legalizer.cpp [hunk 2]
```diff
@@ -0,0 +1,1104 @@
...
+        std::sort(overfilled_bins_vec.begin(), overfilled_bins_vec.end(), [&](LegalizerBinId a, LegalizerBinId b) {
+            return bins_[a].supply.manhatten_norm() < bins_[b].supply.manhatten_norm();
+        });
+        // Get the paths to flow blocks from the overfilled bins to the under
+        // filled bins and flow the blocks.
+        for (LegalizerBinId bin_id : overfilled_bins_vec) {
+            // Get the list of candidate paths based on psi.
+            //  NOTE: The paths are sorted by increasing cost within the
+            //        getPaths method.
+            std::vector<std::vector<LegalizerBinId>> paths = get_paths(bin_id,
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.cpp:1056]: Could put a comment saying "Get multiple candidate paths, each of which is a vector of the legalizerBins traversed."

File: vpr/src/analytical_place/partial_legalizer.cpp [hunk 3]
```diff
@@ -0,0 +1,1104 @@
...
+        // NOTE: Although the supply should always be non-negative, we still
+        //       take the absolute value in the norm for completeness.
+        // TODO: This is a guess. Should investigate other norms.
+        std::vector<LegalizerBinId> overfilled_bins_vec(overfilled_bins_.begin(), overfilled_bins_.end());
+        std::sort(overfilled_bins_vec.begin(), overfilled_bins_vec.end(), [&](LegalizerBinId a, LegalizerBinId b) {
+            return bins_[a].supply.manhatten_norm() < bins_[b].supply.manhatten_norm();
+        });
+        // Get the paths to flow blocks from the overfilled bins to the under
+        // filled bins and flow the blocks.
+        for (LegalizerBinId bin_id : overfilled_bins_vec) {
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.cpp:1052]: would bin_id be better called src_bin_id?

File: vpr/src/analytical_place/partial_legalizer.h [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+ *        VPR.
+ */
+enum class e_partial_legalizer {
+    FLOW_BASED      // Multi-commodity flow-based partial legalizer.
+};
+
+/**
+ * @brief The Partial Legalizer base class
+ *
+ * This provied functionality that all Partial Legalizer will use.
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:42]: Legalizers

File: vpr/src/analytical_place/partial_legalizer.h [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+ * compare different solvers.
+ */
+class PartialLegalizer {
+public:
+    virtual ~PartialLegalizer() {}
+
+    /**
+     * @brief Constructor of the base PartialLegalizer class
+     *
+     * Currently just copies the parameters into the class as member varaibles.
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:55]: Could explain verbosity and its valid range.

File: vpr/src/analytical_place/partial_legalizer.h [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    virtual void legalize(PartialPlacement &p_placement) = 0;
+
+protected:
+
+    /// @brief The APNetlist the legalizer will be legalizing the placement of.
+    ///        It is implied that the netlist is not being modified during
+    ///        global placement.
+    const APNetlist& netlist_;
+
+    /// @brief The verbosity of the log statements within the partial legalizer.
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:83]: Define levels of verbosity.

File: vpr/src/analytical_place/partial_legalizer.h [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+ * This allows a separation between the legalizers and tiles such that a bin may
+ * represent multiple tiles.
+ */
+struct legalizer_bin_tag {};
+typedef vtr::StrongId<legalizer_bin_tag, size_t> LegalizerBinId;
+
+/**
+ * @brief A bin used to contain blocks in the partial legalizer.
+ *
+ * Bins can be thought of as generalized tiles which have a capacity of block
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:105]: block -> blocks

File: vpr/src/analytical_place/partial_legalizer.h [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+ * The capacity, utilization, supply, and demand of the bin are stored as
+ * M-dimensional vectors; where M is the number of models (primitives) in the
+ * device. This allows the bin to quickly know how much of each types of
+ * primitives it can contain and how much of each type it currently contains.
+ */
+struct LegalizerBin {
+    /// @brief The blocks currently contained in this bin.
+    std::unordered_set<APBlockId> contained_blocks;
+
+    /// @brief The maximum mass of each primitive this bin can contain.
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:118]: primitive -> primitive type

File: vpr/src/analytical_place/partial_legalizer.h [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    PrimitiveVector capacity;
+
+    /// @brief The current mass of each primitive this bin contains.
+    PrimitiveVector utilization;
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:122]: primitive -> primitive type

File: vpr/src/analytical_place/partial_legalizer.h [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+ *       less accurate legalizers.
+ */
+class FlowBasedLegalizer : public PartialLegalizer {
+private:
+    /// @brief The maximum number of iterations the legalizer can take. This
+    ///        prevents the legalizer from never converging if there is not
+    ///        enough space to flow blocks.
+    static constexpr size_t max_num_iterations_ = 100;
+
+    /// @brief A list of all the bins in the legalizer.
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/partial_legalizer.h:197]: A vector of all the bins in the legalizer.

File: vpr/src/analytical_place/primitive_vector.h [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+ * efficiently operated upon.
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <unordered_map>
+
+/**
+ * @brief A sparse vector class to store an M-dimensional quantity of primtives
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/primitive_vector.h:17]: primitives

File: vpr/src/analytical_place/primitive_vector.h [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+ * This vector is used to represent the capacity of tiles for different
+ * primitives in a closed form which can be manipulated with math operations.
+ *
+ * This vector is also used to represent the "mass" of AP blocks in primitives,
+ * since an AP block may represent many primitives.
+ *
+ * This vector stores floats since it is expected that, due to some heuristics,
+ * the mass of a block may not be a whole number.
+ *
+ * This class contains usefull operations to operate and compare different
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/primitive_vector.h:29]: useful

File: vpr/src/analytical_place/primitive_vector.h [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    inline bool is_non_negative() const {
+        for (const auto& p : data_) {
+            if (p.second < 0.f)
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * @brief Computes the manhatten (L1) norm of this vector.
```
Comment (vaughnbetz) [inline vpr/src/analytical_place/primitive_vector.h:211]: manhattan

**Review by vaughnbetz — APPROVED** [review]
LGTM

**Comment by vaughnbetz** [comment]
Merge whenever you see fit ...

File: vpr/src/analytical_place/partial_legalizer.cpp [hunk 14]
```diff
@@ -0,0 +1,1104 @@
...
+
+    return neighbor_bin_ids;
+}
+
+/**
+ * @brief Get the center point of a rect
+ *
+ * TODO: Should this be put into vtr_geometry?
+ */
+static inline vtr::Point<double> get_center_of_rect(vtr::Rect<double> rect) {
```
[the remark this answers is not in the document]
Comment (AlexandreSinger) [reply vpr/src/analytical_place/partial_legalizer.cpp:336]: Just looked into it. I do not think this should be in vtr_geometry. The issue I am running into is that getting the center of the rect only makes sense when the units of the rect are floating point. If they are integral it gets a bit wacky. I am going to leave it in her for now.

**Pull request merged**